### ISSUE TITLE
Incompatible code generated with 1.12.6 using function call followed by if statement.

### ISSUE THREAD

**Issue by jo-soft** [jashkenas/coffeescript#4568]
Until version 1.12.5 the code `foo bar if baz` transpiled to `if(baz){foo(bar)}` 
since 1.12.6 this code transpiles to `foo(baz ? bar: void 0)` which breaks old code.

But when i use `mood = greatlyImproved if singing` (from docs) it compiles to `if (singing) {mood = greatlyImproved;}` 
Is this indented? Maybe by 'Bugfixes for incorrect code generated by the ? operator within a termary if statement`? I cant find any Bug report for that.

bg,
Johannes

**Comment by GeoffreyBooth** [comment]
[Are you sure?](http://coffeescript.org/#try:foo%20bar%20if%20baz)

**Comment by connec** [comment]
Perhaps you somehow triggered it being parsed as `foo(bar if baz)`? It would help if you could paste a full example that's reproducible in [Try CoffeeScript](http://coffeescript.org/#try:foo%20bar%20if%20baz) (or a link).

**Comment by jo-soft** [comment]
bigger example of code failing in 1.12.6 bug working in 1.12.5 and lower:
```
test:
  bam: () ->
    fn bar: {
      foo: 123
    } if not condition
```
results in 1.12.6 (example generated by: http://coffeescript.org/#try)
```
({
  test: {
    bam: function() {
      return fn({
        bar: !condition ? {
          foo: 123
        } : void 0
      });
    }
  }
});
```

results in 1.12.5. (example generated by [js2.coffee](http://js2.coffee):

```
({
  test: {
    bam: function() {
      if (!condition) {
        return fn({
          bar: {
            foo: 123
          }
        });
      }
    }
  }
});
```

both examples can also locally be reproduced by transpiling locally

**Comment by vendethiel** [comment]
Paging @helixbass since it might be from #4554 (?).

**Comment by helixbass** [comment]
@vendethiel not a bad guess but #4554 was only merged into `2`

**Comment by GeoffreyBooth** [comment]
It’s from #4532, allowing implicit braces after a `return` or `export default`. Basically, in this code:
```coffee
test:
  bam: () ->
    fn bar: {
      foo: 123
    } if not condition
```
it’s ambiguous what the `if` block should be applied to. “Trailing `if`“ is only meant for a one-line expression; if anything, we should be throwing an error here. How far back is this `if` supposed to look to determine whether something should be output? Just `{ foo: 123 }`? Or `{ bar: { foo: 123 } }`? Or `fn { bar: { foo: 123 } }` or `return fn { bar: { foo: 123 } }`?

If you insist on a trailing `if`, [make it explicit](http://coffeescript.org/#try:test%3A%0A%20%20bam%3A%20()%20-%3E%0A%20%20%20%20(fn%20bar%3A%20%7B%0A%20%20%20%20%20%20foo%3A%20123%0A%20%20%20%20%7D)%20if%20not%20condition):

```coffee
test:
  bam: () ->
    (fn bar: {
      foo: 123
    }) if not condition
```

I don’t think there’s a bug here. Yes the output changed, but it’s as a result of fixing another bug regarding implicit braces and `return`. Basically the output here was relying on that bug to create the desired effect.

**Comment by GeoffreyBooth** [comment]
Actually it's probably due to #4534. But my point still stands.

**Comment by vendethiel** [comment]
I don't think we should be changing such semantics on 1.x.

**Comment by GeoffreyBooth** [comment]
Yeah we shouldn’t have merged #4534 into `master` (assuming that is, in fact, the cause of this). My argument is less valid against that PR than it is against fixing braces after `return`, because braces after `return` is more clearly a bug whereas #4534 adjusts behavior to the way it “should” be.

So I guess the question is what to do about it. We’re keeping #4534 on `2`, so it seems odd to revert it on `master` just to keep it in 2. I was hoping to avoid any more 1.x releases.
